Add: note in the save function documentation to discourage side effects (#15765)
WordPress/gutenberg: docs/designers-developers/developers/block-api/block-edit-save.md

_Note:_ While it is possible to return a string value from `save`, it _will be escaped_. If the string includes HTML markup, the markup will be shown on the front of the site verbatim, not as the equivalent HTML node content. If you must return raw HTML from `save`, use `wp.element.RawHTML`. As the name implies, this is prone to [cross-site scripting](https://en.wikipedia.org/wiki/Cross-site_scripting) and therefore is discouraged in favor of a WordPress Element hierarchy whenever possible.

_Note:_ The save function should be a pure function that depends only on the attributes used to invoke it.

> It can not have any side effect or retrieve information from another source, e.g. it is not possible to use the data module inside it `wp.data.select...`.

Reviewer: I think the mention of "`wp.data.select...`" could help clarify what's referred to, but for what it's worth, we should discourage this use of the window global in blocks _anywhere_, since it cannot account for separate data registries, which will become necessary for e.g. embedded editors (#7119), multiple editors on a page.

> If the function uses external information when the external information changes for the same attribute values, the save result may vary, and when that happens, the block will be considered invalid by the editor.

Reviewer: Some suggestion for phrasing. I assume the intent here is to explain _why_ it can't be done? May also help to direct people to the Validation section, which includes more information about how validation actually works (and to avoid repeated information).

```suggestion
This is because if the external information changes, the block may be flagged as invalid when the post is later edited ([read more about Validation](#validation)).
```

If there is a need to have other information as part of the save, developers can consider one of these two alternatives:
 - Use [dynamic blocks](/docs/designers-developers/developers/tutorials/block-tutorial/creating-dynamic-blocks.md) and dynamically retrieve the required information on the server.

>  - Add another attribute value that contains the needed value that the save function depends on and dynamically update that value in the `edit` component which can depend on external information and contain side effects to update the attribute when the external information changes.

Reviewer: This sentence is a bit difficult to read by its length and lack of punctuation.

Maybe:

```suggestion
 - Store the external value as an attribute which is dynamically updated in the block's `edit` function as changes occur.
```
